components/logins/src/migrate_sqlcipher_db.rs

```diff
@@ -78,12 +78,60 @@ impl Add for MigrationPhaseMetrics {
     }
 }
 
-pub fn migrate_sqlcipher_db_to_plaintext(
+pub fn migrate_logins(
+    path: impl AsRef<Path>,
+    new_encryption_key: &str,
+    sqlcipher_path: impl AsRef<Path>,
+    sqlcipher_key: &str,
+    salt: Option<String>,
+) -> Result<String> {
+    let path = path.as_ref();
+    let sqlcipher_path = sqlcipher_path.as_ref();
+
+    // Need to make sure we have an old db and new one before we start anything
+    let result = match (path.exists(), sqlcipher_path.exists()) {
+        (true, true) => migrate_sqlcipher_db_to_plaintext(
+            &sqlcipher_path,
+            &path,
+            sqlcipher_key,
+            new_encryption_key,
+            salt.as_ref(),
+        ),
+        (false, true) => match sqlcipher_path.to_str() {
+            None => throw!(ErrorKind::InvalidPath(
+                sqlcipher_path.as_os_str().to_owned(),
+            )),
+            Some(s) => throw!(ErrorKind::InvalidDatabaseFile(s.to_string())),
+        },
+        (true, false) => match path.to_str() {
+            None => throw!(ErrorKind::InvalidPath(
+                sqlcipher_path.as_os_str().to_owned(),
+            )),
+            Some(s) => throw!(ErrorKind::InvalidDatabaseFile(s.to_string())),
+        },
+        (false, false) => throw!(ErrorKind::InvalidDatabaseFile(
+            "Both paths failed".to_string(),
+        )),
+    };
+
+    match result {
+        Err(e) => {
+            log::error!("Error migrating sqlcipher DB: {}", e);
+            throw!(ErrorKind::MigrationError(e.to_string()))
+        }
+        Ok(metrics) => {
+            // Do we handle deletion here? or leave it to the client
+            Ok(serde_json::to_string(&metrics)?)
+        }
+    }
+}
+
+fn migrate_sqlcipher_db_to_plaintext(
     old_db_path: impl AsRef<Path>,
     new_db_path: impl AsRef<Path>,
     old_encryption_key: &str,
     new_encryption_key: &str,
-    salt: Option<&str>,
+    salt: Option<&String>,
```

Comment:
Nit: I think `&str` is usually preferable to `&String`.  It's more general since `&String` derefs to it.

Reply:
yeah the &str and String change is quite unfortunate but there is actually no way to define `Option<&str>` in uniffi actually. Since `[ByRef] string?` makes it `&Option<String>` instead of `Option<&str>` and I couldn’t figure how to do an option with referenced string and then once I get rid of the & internally in the option — compiler complains due to unknown size at compile time for str...